### PR TITLE
Add Swift 4 Support

### PR DESCRIPTION
**Pull request by jshier**
This PR adds Swift 4 support by making Alamofire compatible with both Swift 3 and Swift 4. By default it builds using Swift 3, but in environments like CocoaPods or manual integration it can be built in Swift 4 as well. 

This PR also refactors some testing issues and enables background session tests, which seem to be the only place those tests work.

### CODE REVIEW

File: Source/ParameterEncoding.swift [hunk 1]
```diff
@@ -241,8 +241,11 @@ public struct URLEncoding: ParameterEncoding {
             let value = parameters[key]!
             components += queryComponents(fromKey: key, value: value)
         }
-
+        #if swift(>=4.0)
+        return components.map { "\($0.0)=\($0.1)" }.joined(separator: "&")
+        #else
         return components.map { "\($0)=\($1)" }.joined(separator: "&")
+        #endif
```
Comment (cnoon) [inline Source/ParameterEncoding.swift]: To keep formatting consistent across Alamofire, how about?

```swift
    }

#if swift(>=4.0)
    return components.map { "\($0.0)=\($0.1)" }.joined(separator: "&")
#else
    return components.map { "\($0)=\($1)" }.joined(separator: "&")
#endif
```

Where the pound defines are demented one tab.

File: Source/SessionManager.swift [hunk 2]
```diff
@@ -58,10 +58,18 @@ open class SessionManager {
...
+                let (index, languageCode) = enumeratedLanguage
+                let quality = 1.0 - (Double(index) * 0.1)
+            return "\(languageCode);q=\(quality)"
+        }.joined(separator: ", ")
+        #else
         let acceptLanguage = Locale.preferredLanguages.prefix(6).enumerated().map { index, languageCode in
             let quality = 1.0 - (Double(index) * 0.1)
             return "\(languageCode);q=\(quality)"
         }.joined(separator: ", ")
+        #endif
```
Comment (cnoon) [inline Source/SessionManager.swift]: How about?

```swift
    // Accept-Language HTTP Header; see https://tools.ietf.org/html/rfc7231#section-5.3.5
#if swift(>=4.0)
    let acceptLanguage = Locale.preferredLanguages.prefix(6).enumerated().map { enumeratedLanguage in
        let (index, languageCode) = enumeratedLanguage
        let quality = 1.0 - (Double(index) * 0.1)
        return "\(languageCode);q=\(quality)"
    }.joined(separator: ", ")
#else
    let acceptLanguage = Locale.preferredLanguages.prefix(6).enumerated().map { index, languageCode in
        let quality = 1.0 - (Double(index) * 0.1)
        return "\(languageCode);q=\(quality)"
    }.joined(separator: ", ")
#endif
```

File: Tests/ResponseTests.swift [hunk 3]
```diff
@@ -272,13 +272,13 @@ class ResponseJSONTestCase: BaseTestCase {
...
             XCTAssertNotNil(response?.metrics)
         }
 
-        // The `as NSString` cast is necessary due to a compiler bug. See the following rdar for more info.
-        // - https://openradar.appspot.com/radar?id=5517037090635776
-        if let args = (response?.result.value as AnyObject?)?["args" as NSString] as? [String: String] {
+        if
+            let responseDictionary = response?.result.value as? [String : Any],
+            let args = responseDictionary["args"] as? [String : String] {
             XCTAssertEqual(args, ["foo": "bar"], "args should match parameters")
```
Comment (cnoon) [inline Tests/ResponseTests.swift]: How about?

```swift
 if
    let responseDictionary = response?.result.value as? [String: Any],
    let args = responseDictionary["args"] as? [String: String] 
{
    XCTAssertEqual(args, ["foo": "bar"], "args should match parameters")
```

Also note dictionaries are `[String: String]` and not `[String : String]`.

File: Tests/ResponseTests.swift [hunk 4]
```diff
@@ -311,9 +311,9 @@ class ResponseJSONTestCase: BaseTestCase {
             XCTAssertNotNil(response?.metrics)
         }
 
-        // The `as NSString` cast is necessary due to a compiler bug. See the following rdar for more info.
-        // - https://openradar.appspot.com/radar?id=5517037090635776
-        if let form = (response?.result.value as AnyObject?)?["form" as NSString] as? [String: String] {
+        if
+            let responseDictionary = response?.result.value as? [String : Any],
+            let form = responseDictionary["form"] as? [String : String] {
```
Comment (cnoon) [inline Tests/ResponseTests.swift]: Same formatting here...

File: Tests/SessionManagerTests.swift [hunk 5]
```diff
@@ -787,12 +787,12 @@ class SessionManagerConfigurationHeadersTestCase: BaseTestCase {
...
-//    func testThatBackgroundConfigurationHeadersAreSentWithRequest() {
-//        // Given, When, Then
-//        executeAuthorizationHeaderTest(for: .background)
-//    }
+    #if os(macOS)
+    func testThatBackgroundConfigurationHeadersAreSentWithRequest() {
+        // Given, When, Then
+        executeAuthorizationHeaderTest(for: .background)
+    }
+    #endif
```
Comment (cnoon) [inline Tests/SessionManagerTests.swift]: How about?

```swift
#if os(macOS)
    func testThatBackgroundConfigurationHeadersAreSentWithRequest() {
        // Given, When, Then
        executeAuthorizationHeaderTest(for: .background)
    }
#endif
```

File: Tests/SessionManagerTests.swift [hunk 6]
```diff
@@ -840,10 +840,9 @@ class SessionManagerConfigurationHeadersTestCase: BaseTestCase {
             XCTAssertNotNil(response.data, "data should not be nil")
             XCTAssertTrue(response.result.isSuccess, "result should be a success")
 
-            // The `as NSString` cast is necessary due to a compiler bug. See the following rdar for more info.
-            // - https://openradar.appspot.com/radar?id=5517037090635776
             if
-                let headers = (response.result.value as AnyObject?)?["headers" as NSString] as? [String: String],
+                let response = response.result.value as? [String : Any],
```
Comment (cnoon) [inline Tests/SessionManagerTests.swift]: `[String : Any]` -> `[String: Any]`.

File: Tests/URLProtocolTests.swift [hunk 7]
```diff
@@ -92,15 +92,15 @@ class ProxyURLProtocol: URLProtocol {
 
 // MARK: -
 
-extension ProxyURLProtocol: URLSessionDelegate {
+extension ProxyURLProtocol: URLSessionDataDelegate {
```
Comment (cnoon) [inline Tests/URLProtocolTests.swift]: Nice catch!

File: Tests/UploadTests.swift [hunk 8]
```diff
@@ -533,69 +533,66 @@ class UploadMultipartFormDataTestCase: BaseTestCase {
...
+        XCTAssertNotNil(data, "data should not be nil")
+        XCTAssertNil(error, "error should be nil")
+
+        if let streamingFromDisk = streamingFromDisk {
+            XCTAssertTrue(streamingFromDisk, "streaming from disk should be true")
+        } else {
+            XCTFail("streaming from disk should not be nil")
+        }
+    }
+    #endif
```
Comment (cnoon) [inline Tests/UploadTests.swift]: Can we dedent the pound defines?

File: Tests/UploadTests.swift [hunk 9]
```diff
@@ -533,69 +533,66 @@ class UploadMultipartFormDataTestCase: BaseTestCase {
...
+                    response = defaultResponse.response
+                    data = defaultResponse.data
+                    error = defaultResponse.error
+
+                    expectation.fulfill()
+                }
+            case .failure:
+                expectation.fulfill()
+            }
+        })
```
Comment (cnoon) [inline Tests/UploadTests.swift]: How about?

```swift
// When
manager.upload(
    multipartFormData: { multipartFormData in
        multipartFormData.append(french, withName: "french")
        multipartFormData.append(japanese, withName: "japanese")
    }, 
    to: urlString,
    encodingCompletion: { result in
        switch result {
        case let .success(upload, uploadStreamingFromDisk, _):
            streamingFromDisk = uploadStreamingFromDisk

            upload.response { defaultResponse in
                request = defaultResponse.request
                response = defaultResponse.response
                data = defaultResponse.data
                error = defaultResponse.error

                expectation.fulfill()
            }
        case .failure:
            expectation.fulfill()
        }
    }
)
```

File: Source/ParameterEncoding.swift (hunk 1 above)
Comment (jshier) [inline Source/ParameterEncoding.swift]: Addressed in [a0337](https://github.com/Alamofire/Alamofire/pull/2163/commits/03dbf047a5fd956e850a135efbf651c390ea0337).

File: Source/SessionManager.swift (hunk 2 above)
Comment (jshier) [inline Source/SessionManager.swift]: Addressed in [a0337](https://github.com/Alamofire/Alamofire/pull/2163/commits/03dbf047a5fd956e850a135efbf651c390ea0337).

File: Tests/ResponseTests.swift (hunk 3 above)
Comment (jshier) [inline Tests/ResponseTests.swift]: Addressed in [a0337](https://github.com/Alamofire/Alamofire/pull/2163/commits/03dbf047a5fd956e850a135efbf651c390ea0337).

File: Tests/ResponseTests.swift (hunk 4 above)
Comment (jshier) [inline Tests/ResponseTests.swift]: Addressed in [a0337](https://github.com/Alamofire/Alamofire/pull/2163/commits/03dbf047a5fd956e850a135efbf651c390ea0337).

File: Tests/SessionManagerTests.swift (hunk 5 above)
Comment (jshier) [inline Tests/SessionManagerTests.swift]: Addressed in [a0337](https://github.com/Alamofire/Alamofire/pull/2163/commits/03dbf047a5fd956e850a135efbf651c390ea0337).

File: Tests/SessionManagerTests.swift (hunk 6 above)
Comment (jshier) [inline Tests/SessionManagerTests.swift]: Addressed in [a0337](https://github.com/Alamofire/Alamofire/pull/2163/commits/03dbf047a5fd956e850a135efbf651c390ea0337).

File: Tests/URLProtocolTests.swift (hunk 7 above)
Comment (jshier) [inline Tests/URLProtocolTests.swift]: Addressed in [a0337](https://github.com/Alamofire/Alamofire/pull/2163/commits/03dbf047a5fd956e850a135efbf651c390ea0337).

File: Tests/UploadTests.swift (hunk 9 above)
Comment (jshier) [inline Tests/UploadTests.swift]: Addressed in [a0337](https://github.com/Alamofire/Alamofire/pull/2163/commits/03dbf047a5fd956e850a135efbf651c390ea0337).

File: Tests/UploadTests.swift (hunk 8 above)
Comment (jshier) [inline Tests/UploadTests.swift]: Addressed in [a0337](https://github.com/Alamofire/Alamofire/pull/2163/commits/03dbf047a5fd956e850a135efbf651c390ea0337).

**Pull request merged**